### ISSUE TITLE
[Feature Request] Improvement to correlation functionality in the case of binary data

### ISSUE THREAD

**Issue by 411an13** [jolespin/ensemble_networkx#3]
It seems like there is currently no way to robustly do correlations on binary data (e.g. correlations between two dichotomous categorical variables).
I'll try to implement the Matthews Correlation Coefficient (also known as the Phi Coefficient) in the **EnsembleAssociationNetwork class** to resolve this.

**Comment by jolespin** [comment]
Brings up a great point!

We are going to want to the input handle both NumPy arrays and Pandas DataFrames (similar to this but `pairwise_mcc` will be way less complicated: https://github.com/jolespin/compositional/blob/cd8949a7e78f029c6e88e7e7432b6b3776842ff3/compositional/compositional.py#L80C11-L80C11)

Let's get a for-loop working based on actual formula  that matches the Sklearn output.  

Here's the formula: https://www.statology.org/matthews-correlation-coefficient-python/

Here's a few implementations: 
* https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.contingency.association.html
* https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
* https://towardsdatascience.com/correlation-when-pearsons-r-is-not-enough-aded72308635

We don't want to add dependencies like Sklearn unless we absolutely need them and this one seems simple enough to code up.  

Once we get a for-loop version that works exactly as expected, then we can work on vectorizing the computations so we won't use any for-loop and we can let linear algebra do it's magic.

**Comment by 411an13** [comment]
# Summary of the new function
We created a new function, `pairwise_mcc`, that can accept NumPy arrays and Pandas DataFrames as input. 
It returns a pairwise implementation of the Matthews correlation coefficient, which is a statistically valid metric for the association of dichotomous variables.

### Draft version of the function that utilized for-loops to calculate the MCC values.
```
def pairwise_mcc(X:pd.DataFrame, check=True):
    """
    # Description
    Returns a correlation table containing Matthews correlation coefficients for a given matrix (np.array or pd.dataframe) of binary, categorical variables.
    This implementation was created as an alternative to Scikit-Learn's implementation as a measure of dependency reduction.

    # Scikit-Learn implementation of MCC:
    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
    
    # Parameters
        * X:
            - NumPy array or Pandas dataframe
    """
    # Checks
    if check:
        n_dimensions = len(X.shape)
        assert n_dimensions in {2}, "`X` must be 2D"
        assert np.all(X == X.astype(bool)), "`X` must be either dtype boolean or integers[0,1]"

    # Convert input data to a NumPy array
    # index=None
    components=None
    if isinstance(X, pd.DataFrame):
        # index = X.index
        features = X.columns
        X = X.values
        
    X = X.astype(bool)
    
    # Shape of matrix
    n,m = X.shape

    # Initialize the resulting matrix to store MCC values
    output = np.zeros((m, m), dtype=float)
    
    # Calculate pairwise MCC values
    for i in range(m):
        u = X[:, i]
        for j in range(i+1, m):  # We only need to calculate the upper triangular part
            v = X[:, j]
            n11 = np.sum((u == 1) & (v == 1))
            n10 = np.sum((u == 1) & (v == 0))
            n01 = np.sum((u == 0) & (v == 1))
            n00 = np.sum((u == 0) & (v == 0))
            
            mcc = (n11 * n00 - n10 * n01) / np.sqrt((n11 + n10) * (n11 + n01) * (n00 + n10) * (n00 + n01))
            
            output[i, j] = output[j, i] = mcc
    np.fill_diagonal(output, 1.0)
    
    # Return the result as a DataFrame if the input was a DataFrame
    if features is not None:
        output = pd.DataFrame(output, index=features, columns=features)
        
    return output
```

#### Sklearn-based version

```
from sklearn.metrics import matthews_corrcoef


def pairwise_mcc_sklearn(data):
    """
    # Description
    Returns a correlation table containing Matthews correlation coefficients for a given matrix (np.array or pd.dataframe) of binary, categorical variables.
    This implementation was created as an alternative to Scikit-Learn's implementation as a measure of dependency reduction.

    # Scikit-Learn implementation of MCC:
    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
    
    # Parameters
        * X:
            - NumPy array or Pandas dataframe
    """
    # MODIFY THIS CODE:
    # Convert input data to a NumPy array
    if isinstance(data, pd.DataFrame):
        data = data.values

    # Calculate the number of columns
    num_columns = data.shape[1]

    # Initialize the resulting matrix to store MCC values
    mcc_matrix = np.zeros((num_columns, num_columns))

    # Calculate pairwise MCC values
    for i in range(num_columns):
        for j in range(i, num_columns):  # We only need to calculate the upper triangular part
            y_true = data[:, i]
            y_pred = data[:, j]
            mcc = matthews_corrcoef(y_true, y_pred)

            mcc_matrix[i, j] = mcc
            mcc_matrix[j, i] = mcc  # Since MCC is symmetric, we fill both (i, j) and (j, i)

    # Return the result as a DataFrame if the input was a DataFrame
    if isinstance(data, pd.DataFrame):
        mcc_df = pd.DataFrame(mcc_matrix, columns=data.columns, index=data.columns)
        return mcc_df
    else:
        return mcc_matrix
```

### New version that replaces the for-loops with vectorization
```
def pairwise_mcc_vectorized(X:pd.DataFrame, check=True):
    """
    # Description
    Returns a correlation table containing Matthews correlation coefficients for a given matrix (np.array or pd.dataframe) of binary, categorical variables.
    This implementation was created as an alternative to Scikit-Learn's implementation as a measure of dependency reduction.

    # Scikit-Learn implementation of MCC:
    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html
    
    # Parameters
        * X:
            - NumPy array or Pandas dataframe
    Output: 
        pd.DataFrame or np.array of pairwise MCC values
    """
    # Checks
    if check:
        n_dimensions = len(X.shape)
        assert n_dimensions in {2}, "`X` must be 2D"
        assert np.all(X == X.astype(bool)), "`X` must be either dtype boolean or integers[0,1]"

    # Convert input data to a NumPy array
    # index=None
    components=None
    if isinstance(X, pd.DataFrame):
        # index = X.index
        features = X.columns
        X = X.values
        
    X = X.astype(bool)
    
    # Shape of matrix
    n,m = X.shape

    # Calculate pairwise MCC values
    N11 = (X[:, None] & X[:, :, None]).sum(axis=0)
    N10 = (X[:, None] & ~X[:, :, None]).sum(axis=0)
    N01 = (~X[:, None] & X[:, :, None]).sum(axis=0)
    N00 = (~X[:, None] & ~X[:, :, None]).sum(axis=0)
    denominator = np.sqrt((N11 + N10) * (N11 + N01) * (N00 + N10) * (N00 + N01))
    denominator[denominator == 0] = 1  # Handle division by zero case

    output = (N11 * N00 - N10 * N01) / denominator

    # Fill the lower triangular part with the symmetric values
    output[np.tril_indices(m, k=-1)] = output.T[np.tril_indices(m, k=-1)]

    # Set diagonal elements to 1.0
    np.fill_diagonal(output, 1.0)

    # Return the result as a DataFrame if the input was a DataFrame
    if features is not None:
        output = pd.DataFrame(output, index=features, columns=features)
        
    return output
```

### Benchmarking
Note: In the code block below, `pairwise_mcc` refers to an old version of `pairwise_mcc_vectorized`, the latter of which will be renamed to 'pairwise_mcc' and pushed to the repo. `pairwise_mcc_scikit` refers to a version of the old `pairwise_mcc` function (with for-loops) that utilizes Scikit-Learn's implementation of MCC, available at https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html.
```
df = pd.DataFrame(np.random.RandomState(42).randint(0,2,size=(1000,100)))
print(np.allclose(pairwise_mcc(df), pairwise_mcc_scikit(df)))
print(np.allclose(pairwise_mcc(df), pairwise_mcc_vectorized(df)))

%timeit pairwise_mcc(df)
%timeit pairwise_mcc_scikit(df)
%timeit pairwise_mcc_vectorized(df)
```
The output from running the previous code block:
_True
True
248 ms ± 15.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
5.35 s ± 323 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
43.4 ms ± 3.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)_

### **`pairwise_mcc` will be added to the `EnsembleAssociationNetwork` class within `ensemble_networkx`**

**Comment by jolespin** [comment]
This all looks great.  The next steps will be to add this to `ensemble_networkx.py` and the function name to `__init__.py`.  Once we get that properly integrated, we can do the following: 
* Add "mcc" as a metric option to the following classes: 
    * EnsembleAssociationNetwork
    * SampleSpecificPerturbationNetwork
    * DifferentialEnsembleAssociationNetwork
Eventually we will want to implement some unit tests but right now, we can just test manually.

**Comment by 411an13** [comment]
Made the pull request. It should have 2 commits.

**Comment by 411an13** [comment]
The 1st pull request I made only had 1 of the commits. The 2nd one, called "Pairwise mcc update", has both as well as test code and the results from testing the new function.

**Comment by jolespin** [comment]
Added in https://github.com/jolespin/ensemble_networkx/releases/tag/v2023.7.20
